hphp/runtime/vm/extern-compiler.cpp

```diff
@@ -181,6 +181,94 @@ struct CompilerOptions {
 
 constexpr int kInvalidPid = -1;
 
+struct Pipe final {
+  explicit Pipe(bool nonblocking = false) {
+    int flags = O_CLOEXEC;
+    if (nonblocking) flags |= O_NONBLOCK;
+
+    if (pipe2(fds, flags) == -1) throwErrno("unable to open pipe");
+  }
+  ~Pipe() {
+    if (fds[0] != -1) close(fds[0]);
+    if (fds[1] != -1) close(fds[1]);
+  }
+  FILE* detach(const char* mode) {
+    auto ret = fdopen(fds[*mode == 'r' ? 0 : 1], mode);
+    if (!ret) throwErrno("unable to fdopen pipe");
+    close(fds[*mode == 'r' ? 1 : 0]);
+    fds[0] = fds[1] = -1;
+    return ret;
+  }
+  int remoteIn() const { return fds[0]; }
+  int remoteOut() const { return fds[1]; }
+  int fds[2];
+};
+
+std::string readline(FILE* f);
+
+struct LogThread {
+  LogThread(int pid, FILE* file)
+    : m_pid(pid), m_file(file), m_signalPipe(std::make_unique<Pipe>(true))
```

Comment:
from @paulbiss : Does this need to be a pointer?

Reply:
No, I can convert it to an fd and just keep track of that.

Reply:
Oops. Spoke too soon. It gets passed `readline`, which is expecting `FILE*`.

Reply:
And I don't think it's safe to pass non-pointer FILE around, if that's what you were asking for.

Reply:
Sorry, to clarify, does `Pipe` need to be a pointer? The FILE* is fine

Reply:
I chose to store it as a pointer so that it can be destroyed, closing the pipe's file descriptors, when the thread is stopped (and destroyed). Strictly speaking, no it doesn't need to be a pointer, though I think I'd want to add a close method to Pipe to close the fds in LogThread::stop.